src/qibolab/instruments/rfsoc.py

```diff
@@ -211,43 +307,50 @@ def play(
             qibolab results objects
         """
 
-        if execution_parameters.acquisition_type is AcquisitionType.RAW:
-            raise NotImplementedError("Raw data acquisition is not supported")
-        if execution_parameters.fast_reset:
-            raise NotImplementedError("Fast reset is not supported")
-        # if new value are passed, they are updated in the config obj
-        if execution_parameters.nshots is not None:
-            self.cfg.reps = execution_parameters.nshots
-        if execution_parameters.relaxation_time is not None:
-            self.cfg.repetition_duration = execution_parameters.relaxation_time
+        self.validate_sequence_and_update_cfg(sequence, execution_parameters)
 
         if execution_parameters.acquisition_type is AcquisitionType.DISCRIMINATION:
             average = False
         else:
             average = execution_parameters.averaging_mode is AveragingMode.CYCLIC
 
-        toti, totq = self._execute_pulse_sequence(self.cfg, sequence, qubits, len(sequence.ro_pulses), average)
+        toti, totq = self._execute_pulse_sequence(sequence, qubits, average)
 
         results = {}
-        adcs = np.unique([qubits[p.qubit].feedback.ports[0][1] for p in sequence.ro_pulses])
-        for j in range(len(adcs)):
-            for i, ro_pulse in enumerate(sequence.ro_pulses):
+        adc_chs = np.unique([qubits[p.qubit].feedback.ports[0][1] for p in sequence.ro_pulses])
+
+        for j in range(len(adc_chs)):
```

Comment:
Usually iterations on `range(len(...))` can be replaced with iterations on `...`, or some other associated iterator.

But I haven't been through the whole thing, so here I don't know exactly how (yet).

Reply:
I think he's only using `j` inside, and not `adc_chs[j]`, so that's probably why he opted to do it this way.

Reply:
I improved it a bit with enumerate